Init : 엔티티 (DB 테이블) 작성

adorableco opened the pull request:
## **PR**

### ✨ 작업 내용

- `Member`, `Question`, `QuestionSet`, `Tag`, `WorkBook` 엔티티 작성

---

### ✨ 참고 사항

- 코틀린 코드 내에서 모든 엔티티의 pk 는 `id` 로 명명하고 칼럼명을 `엔티티_id` 로 지정함


<img width="483" alt="스크린샷 2024-07-01 오후 9 20 56" src="https://github.com/swm-standard/Phote_BE/assets/90397541/90f76d19-f55f-4114-a588-56964def893f">




---

### ⏰ 현재 버그

x

---

### ✏ Git Close #1

adorableco committed:
Init : 엔티티 (DB 테이블) 작성

- `Member`, `Question`, `QuestionSet`, `Tag`, `WorkBook` 엔티티 작성

adorableco committed:
Init : Question 엔티티 `category` 필드 삭제

- `tag` 가 있으므로 `category` 가 불필요함

adorableco committed:
Init : tag-question N:M 해결용 중간 테이블 추가

- `QUESTION` 테이블의 `category` 재추가

RinRinPARK commented on src/main/kotlin/com/swm_standard/phote/entity/Question.kt:
```diff
@@ -0,0 +1,39 @@
+    @OneToMany(mappedBy = "question", cascade = [(CascadeType.REMOVE)])
+    val questionSet: Set<QuestionSet>?,
+
+    @JoinColumn(name = "tag_id", nullable = true)
+    @OneToMany
+    val tags: Set<QuestionTag>,
```
question이 삭제되면 tag와의 연관(?)은 삭제되어야 하니까 `cascade = [(CascadeType.REMOVE)]`옵션 걸어주는게 좋을 것 같습니다!

adorableco replied:
이것도 약간 헷갈리는게 question도 삭제 시 `deleted_at` 필드에 날짜 표기하는걸로 처리할텐데, 이때 `question_tag` 랑 `tag` 로 `delete_at` 표기로 처리해야하지 않을까요?

RinRinPARK replied:
아 그르넹 Question이 물리적으로 삭제될 일이 없군여 ㅋㅋ헷 세연님 말이 맞는 것 같슴다

adorableco replied:
녭녭 리뷰해주신거 반영해둘게요 ~ ~ ~

RinRinPARK commented on src/main/kotlin/com/swm_standard/phote/entity/WorkBook.kt:
```diff
@@ -0,0 +1,30 @@
+
+    val description: String?,
+
+    @ManyToOne(cascade = [(CascadeType.REMOVE)])
+    @JoinColumn(name = "member_id")
+    val member: Member,
```
이 코드는 workbook이 삭제되면 유저가 같이 삭제되어서 이부분 cascade는 없애야 될 것 같습니다!

adorableco replied:
헉 반영하겠습니다~~

RinRinPARK commented on src/main/kotlin/com/swm_standard/phote/entity/QuestionSet.kt:
```diff
@@ -0,0 +1,22 @@
+    val workbook: WorkBook,
+
+    val createdAt: LocalDateTime,
+    val deletedAt: LocalDateTime?,
```
음 지금 question이랑 workbook 둘 다 question set에 대해 cascade걸려 있어서 만약 얘네가 삭제되면 question set도 자동으로 hard delete 되고 있는 것 같습니다. 그러면 delete_at을 없애도 되지 않을까엽??

adorableco replied:
삭제는 **deletedAt 으로 삭제 일시를 표기하는걸**로 간주하고 있긴 한데 일단 cascade 옵션도 추가하기는 했어욥 일단 deletedAt  필드를 사용하니깐 cascade 옵션은 그냥 제거하는게 나을까요?

RinRinPARK replied:
예옙 둘 다 물리적으로 삭제될 일 없을 것 같아서 cascade 없애도 될 것 같슴돠

RinRinPARK reviewed (commented):
😺👍

adorableco committed:
Fix : 엔티티 cascade 옵션 삭제

- 엔티티명 수정 : `WorkBook` -> `Workbook`
- `Workbook` 엔티티의 `member` 필드 cascade 옵션 삭제
- `Workbook`, `question` 엔티티의 `questionSet` 필드 cascade 옵션 삭제

adorableco committed:
Fix: 엔티티 보완

- UUID 대체키 추가
- `Member` 엔티티 로그인 용 속성 (name, email, image, provider) 추가
- `Question` 과 `Tag` 의 연관관계를 N:M 에서 1:N 으로 변경
- `Workbook` 엔티티에 문제 개수, 이모지 문자열, 수정일자 추가
- `Question`엔티티에 객관식 용 options, 수정일 추가
- `QuestionSet` 엔티티에 문제 순서 추가

related to: #1
